stdlib/public/core/AnyHashable.swift

```diff
@@ -39,18 +39,36 @@ public protocol _HasCustomAnyHashableRepresentation {
 
 @usableFromInline // FIXME(sil-serialize-all)
 internal protocol _AnyHashableBox {
-  func _unbox<T : Hashable>() -> T?
+  var _canonicalBox: _AnyHashableBox { get }
 
   /// Determine whether values in the boxes are equivalent.
   ///
+  /// - Precondition: `self` and `box` are in canonical form.
   /// - Returns: `nil` to indicate that the boxes store different types, so
   ///   no comparison is possible. Otherwise, contains the result of `==`.
-  func _isEqual(to: _AnyHashableBox) -> Bool?
+  func _isEqual(to box: _AnyHashableBox) -> Bool?
   var _hashValue: Int { get }
   func _hash(into hasher: inout Hasher)
+  func _rawHashValue(_seed: (UInt64, UInt64)) -> Int
 
   var _base: Any { get }
+  func _unbox<T: Hashable>() -> T?
   func _downCastConditional<T>(into result: UnsafeMutablePointer<T>) -> Bool
+
+  func _asSet() -> Set<AnyHashable>?
+  func _asDictionary() -> Dictionary<AnyHashable, AnyHashable>?
```

Comment:
It's a little strange to have to enumerate the `Set` and `Dictionary` types here, but it seems like you're doing it as a way to get right at the canonical `Set` or `Dictionary`. It's a small thing, but a name like `_asCanonicalSet` might make it more obvious why callers wouldn't use `_downCastConditional` instead. (I expand on this comment a bit below).

I briefly got concerned about compatibility going forward, if we were to add another special type like `Set` or `Dictionary`, but we could add the corresponding `_asFoo` in the same way (with a default implementation) at the same time. So, I'm not concerned about this.

Reply:
I intend to remove `@usableFromInline` from this protocol, so we should be able to change it however we like!

I *think* we could eliminate these two unsightly requirements by adapting `_ArrayAnyHashableProtocol` to sets/dictionaries:

```swift
internal protocol _SetAnyHashableProtocol: _AnyHashableBox {
  var count: Int { get }
  func contains(_ member: AnyHashable) -> Bool
}

internal struct _SetAnyHashableBox<Element: Hashable>: _SetAnyHashableProtocol {
  let _base: Set<Element>

  func contains(_ member: AnyHashable) -> Bool {
    guard let member = member as? Element else { return false }
    return _base.contains(member)
  }
  func _isEqual(to other: _AnyHashableBox) -> Bool? {
    guard let other = other as? _SetAnyHashableProtocol else { return nil }
    guard let _base.count == other.count else { return false }
    for member in _base {
      guess other.contains(member as AnyHashable) else { return false }
    }
    return true
  }
  // etc.
}
```

This would clean up the interface and also eliminate the need for storing a full copy of the set/dictionary in `_canonical`. However, it would require downcasting from AnyHashable to any ==-compatible type, which isn't rock solid yet.

Reply:
I realized that we can use `_canonicalBox` to force all sets/dictionaries into a single known box type, so we can get rid of `_asSet()/_asDictionary()` without relying on downcasts. 🎉

Implemented in 2f4ad7982de9f1e5b29f7d8ee275ada74e0b0809

Reply:
Ahhh, that's a good solution!